pkgs/misc/uboot/hardkernel-firmware.nix

```diff
@@ -0,0 +1,136 @@
+{ stdenv
+, lib
+, fetchpatch
+, fetchFromGitHub
+, buildPackages
+, pkgsCross
+}:
+
+let
+  buildHardkernelFirmware = {
+    version ? null
+    , src ? null
+    , name ? ""
+    , filesToInstall
+    , installDir ? "$out"
+    , defconfig
+    , extraMeta ? {}
+    , ... } @ args: stdenv.mkDerivation ({
+      pname = "uboot-hardkernel-firmware-${name}";
+
+      nativeBuildInputs = [
+        buildPackages.git
+        buildPackages.hostname
+        pkgsCross.arm-embedded.stdenv.cc
+      ];
+
+      depsBuildBuild = [
+        buildPackages.gcc49
+      ] ++ lib.optional (stdenv.buildPlatform != stdenv.hostPlatform) buildPackages.stdenv.cc
+        ++ lib.optional (!stdenv.isAarch64) pkgsCross.aarch64-multiplatform.buildPackages.gcc49;
+
+      postPatch = ''
+        substituteInPlace Makefile --replace "/bin/pwd" "pwd"
+      '';
+
+      makeFlags = [
+        "CROSS_COMPILE=${stdenv.cc.targetPrefix}"
+        "CROSS_COMPILE_32=${pkgsCross.arm-embedded.stdenv.cc.targetPrefix}"
+        "${defconfig}" "bl301.bin"
+      ]
+      ++ lib.optional (!stdenv.isAarch64) "CROSS_COMPILE=${pkgsCross.aarch64-multiplatform.stdenv.cc.targetPrefix}";
+
+      installPhase = ''
+        mkdir -p ${installDir}
+        cp ${lib.concatStringsSep " " filesToInstall} ${installDir}
+      '';
+
+      meta = with lib; {
+        homepage = "https://www.hardkernel.com/";
+        description = "Das U-Boot from Hardkernel with Odroid embedded devices firmware and support";
+        license = licenses.unfreeRedistributableFirmware;
+        maintainers = with maintainers; [ aarapov ];
+      } // extraMeta;
+    } // removeAttrs args [ "extraMeta" ]);
+in {
+  inherit buildHardkernelFirmware;
+
+  firmwareOdroidC2 = buildHardkernelFirmware {
+    defconfig = "odroidc2_config";
+    name = "firmware-odroid-c2";
+    version = "2015.01";
+    src = fetchFromGitHub {
+      owner = "hardkernel";
+      repo = "u-boot";
+      rev = "fac4d2da0a1b61dfdeaca0034a45151ff5983fb8";
+      sha256 = "09s0y69ilrwnvqi1g11axsnhylq8kfljwqxdfjifa227mi0kzq37";
+    };
+
+    # https://wiki.odroid.com/odroid-c2/software/building_u-boot
+    prePatch = ''
```

Comment:
Why are the patches not in patches/patchPhase?

Reply:
AFAIK, we have to use either prePatch or postPatch phase here so that we can explicitly call patch util to patch the sources (instead of using builtin patch func, which is not capable to move and create new files).

@SuperSandro2000, I also might not be aware of some other elegant way to apply this patches here.

Reply:
Then postPatch should be used because prePatch breaks patches put in patches.

Reply:
Have you tried using `patches = [ ... ]`? I don't think it has any limitations regarding moving/creating files because it ends up calling the patch command in pretty much the same way you are, see: https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/setup.sh#L948

Reply:
@lopsided98, indeed,... I do remember it didn't work for me back in October and don't remember why. Now it just works. I guess I used it somehow differently.
@SuperSandro2000, even after switching to `patches = [ ... ]` there are few edits in Makefiles. I feel like preBuild would be a better place to make changes in Makefiles, let me know if you think otherwise.